core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AddExchanges.java

```diff
@@ -1103,7 +1107,7 @@ private Partitioning selectUnionPartitioning(UnionNode node, PreferredProperties
             // Try planning the children to see if any of them naturally produce a partitioning (for now, just select the first)
             boolean nullsAndAnyReplicated = parentPreference.isNullsAndAnyReplicated();
             for (int sourceIndex = 0; sourceIndex < node.getSources().size(); sourceIndex++) {
-                PreferredProperties.PartitioningProperties childPartitioning = parentPreference.translate(outputToInputTranslator(node, sourceIndex)).get();
+                PreferredProperties.PartitioningProperties childPartitioning = parentPreference.translate(outputToInputTranslator(node, sourceIndex)).orElse(partitioned(ImmutableSet.of()));
```

Comment:
The fact that parentPreference includes a column that does not exist in the child seems like a mistake. This is masking around the underlying problem.

Reply:
Indeed. We just didn't handle AssignUniqueId correctly.